### PR TITLE
Move sObject Refresh command to apex module and improve telemetry

### PR DESCRIPTION
**Pull request by brpowell**
### What does this PR do?

The command code has been moved to the apex module, and it now kicks off the process once the language server has finished indexing the classes. This is to minimize the language server initialization clashing with pulling down all of the classes. The setting to enable or disable refresh on start is still present, but it is now under the apex configuration. Send additional telemetry data about the refresh process.

### What issues does this PR fix or reference?

W-6017035

### CODE REVIEW

File: packages/salesforcedx-vscode-apex/src/messages/i18n.ts [hunk 1]
```diff
@@ -49,5 +50,7 @@ export const messages = {
     'It looks like this file has been updated. To update your code coverage numbers, run the tests in this file.',
   colorizer_no_code_coverage_current_file:
     'No code coverage information was found for this file. Set "salesforcedx-vscode-core.retrieve-test-code-coverage": true in your user or workspace settings. Then, run Apex tests that include methods in this file. You can run tests from the Apex Tests sidebar or using the Run Tests or Run All Tests code lens within the file.',
-  colorizer_statusbar_hover_text: 'Highlight Apex Code Coverage'
+  colorizer_statusbar_hover_text: 'Highlight Apex Code Coverage',
+  force_sobjects_no_refresh_if_already_active_error_text:
+    'A refresh of your sObject definitions is already underway. If you need to restart the process, cancel the running task.'
```
Comment (ruthemmanuelle) [inline packages/salesforcedx-vscode-apex/src/messages/i18n.ts]: 👍

File: packages/salesforcedx-vscode-core/src/index.ts [hunk 2]
```diff
@@ -425,29 +418,25 @@ export async function activate(context: vscode.ExtensionContext) {
...
-    );
-  }
-
   const api: any = {
-    ProgressNotification,
+    channelService,
     CompositeParametersGatherer,
     EmptyParametersGatherer,
+    getDefaultUsernameOrAlias,
+    getRootWorkspacePath,
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/index.ts]: I don't think we should expose `getRootWorkspacePath` to other extensions. It is a nice wrapper to the vscode workspace api but it does not add anything to justify other extensions using it.

**Comment by praksb** [comment]
@brpowell Is it possible to make the change to fetch the files before we start indexing in LSP? Otherwise this change would be a potential source of performance issue since I think an sobject fetch would cause us to reinitialize again. So we would have to run index once at startup and again on an sobject fetch.

**Comment by brpowell** [comment]
> @brpowell Is it possible to make the change to fetch the files before we start indexing in LSP? Otherwise this change would be a potential source of performance issue since I think an sobject fetch would cause us to reinitialize again. So we would have to run index once at startup and again on an sobject fetch.

You mean wait for sObject refresh to be done and then initialize the LSP? Or kick it off in parallel? If it's the former I'd say that might be a burden for the user to have to wait for refresh to finish to start using lsp features.

**Comment by lcampos** [comment]
@praksb @brpowell indexing and recompilations are good things to look out for with these changes. Keep in mind that we don't know how long the fetching of sobjects will take since that'll depend on the org's shape. You can look at existing telemetry to get an idea, but it could be that indexing local code finishes before the sobject fetch call.

**Comment by praksb** [comment]
@brpowell I was thinking of the former (wait for sobject refresh to be done not run in parallel). But, I do see your point that the refresh might take a while. We can check this in as is and wait for customer feedback. The possible perf impact should also be minimal since we do a refresh only once now and that too only when the folder didn't exist.

> > @brpowell Is it possible to make the change to fetch the files before we start indexing in LSP? Otherwise this change would be a potential source of performance issue since I think an sobject fetch would cause us to reinitialize again. So we would have to run index once at startup and again on an sobject fetch.
> 
> You mean wait for sObject refresh to be done and then initialize the LSP? Or kick it off in parallel? If it's the former I'd say that might be a burden for the user to have to wait for refresh to finish to start using lsp features.

File: packages/salesforcedx-vscode-apex/package.json [hunk 3]
```diff
@@ -225,6 +226,13 @@
           ],
           "default": false,
           "description": "%apex_semantic_errors_description%"
+        },
+        "salesforcedx-vscode-apex.enable-sobject-refresh-on-startup": {
+          "type": [
+            "boolean"
+          ],
+          "default": true,
+          "description": "%enable_sobject_refresh_on_startup_description%"
```
Comment (lcampos) [inline packages/salesforcedx-vscode-apex/package.json]: The `sfdx.force.internal.refreshsobjects` commands from the core extension's `package.json` need to be moved to this `package.json`

File: packages/salesforcedx-sobjects-faux-generator/src/generator/fauxClassGenerator.ts [hunk 4]
```diff
@@ -24,6 +24,21 @@ export interface CancellationToken {
   isCancellationRequested: boolean;
```
Comment (brpowell) [inline packages/salesforcedx-sobjects-faux-generator/src/generator/fauxClassGenerator.ts]: Changes in this file are for improving telemetry data around this command

File: packages/salesforcedx-vscode-apex/package.json [hunk 5]
```diff
@@ -225,6 +230,13 @@
           ],
           "default": false,
           "description": "%apex_semantic_errors_description%"
+        },
+        "salesforcedx-vscode-apex.enable-sobject-refresh-on-startup": {
+          "type": [
+            "boolean"
+          ],
+          "default": false,
```
Comment (brpowell) [inline packages/salesforcedx-vscode-apex/package.json]: As we discussed this morning, the default behavior will now be off. The PR is mostly focused on moving the code to the apex extension and improving telemetry data now.

File: packages/salesforcedx-vscode-apex/src/commands/forceGenerateFauxClasses.ts [hunk 6]
```diff
@@ -112,18 +111,30 @@ export class ForceGenerateFauxClassesExecutor extends SfdxCommandletExecutor<{}>
...
-      console.log('Generate error ' + e);
+      const result = await gen.generate(
+        vscode.workspace.workspaceFolders![0].uri.fsPath,
+        SObjectCategory.ALL,
+        refreshSource
+      );
+      console.log('Generate success ' + result.data);
+      this.logMetric(commandName, startTime, result.data);
+    } catch (result) {
+      console.log('Generate error ' + result.error);
```
Comment (brpowell) [inline packages/salesforcedx-vscode-apex/src/commands/forceGenerateFauxClasses.ts]: Now log errors and successes with the number of objects fetched - if the command got far enough to collect that information.

File: packages/salesforcedx-vscode-apex/src/index.ts [hunk 7]
```diff
@@ -59,14 +67,10 @@ export async function activate(context: vscode.ExtensionContext) {
   }
 
   // Telemetry
-  if (sfdxCoreExtension && sfdxCoreExtension.exports) {
-    sfdxCoreExtension.exports.telemetryService.showTelemetryMessage();
```
Comment (brpowell) [inline packages/salesforcedx-vscode-apex/src/index.ts]: @lcampos removed the telemetry message based on our discussion

File: packages/salesforcedx-vscode-apex/package.json [hunk 8]
```diff
@@ -205,6 +206,10 @@
       {
         "command": "sfdx.force.apex.test.last.method.run",
         "title": "%force_apex_test_last_method_run_text%"
+      },
+      {
+        "command": "sfdx.force.internal.refreshsobjects",
```
Comment (lcampos) [inline packages/salesforcedx-vscode-apex/package.json]: You are missing porting the command under `commandPalette`.

File: packages/salesforcedx-vscode-apex/src/index.ts [hunk 9]
```diff
@@ -80,6 +84,17 @@ export async function activate(context: vscode.ExtensionContext) {
       if (languageClient) {
         languageClient.onNotification('indexer/done', async () => {
           LanguageClientUtils.indexing = false;
+
+          // Refresh SObject definitions if there aren't any faux classes
+          const sobjectRefreshStartup: boolean = vscode.workspace
+            .getConfiguration(SFDX_APEX_CONFIGURATION_NAME)
+            .get<boolean>(ENABLE_SOBJECT_REFRESH_ON_STARTUP, true);
```
Comment (lcampos) [inline packages/salesforcedx-vscode-apex/src/index.ts]: The default should be `false`

Comment (brpowell) [reply packages/salesforcedx-vscode-apex/src/index.ts]: I'll change that for consistency. FYI I recently discovered this value doesn't actually matter here, the default value is read from the package.json.

File: packages/salesforcedx-vscode-apex/src/telemetry/telemetry.ts [hunk 10]
```diff
@@ -79,6 +87,29 @@ export class TelemetryService {
     }
   }
 
+  public sendErrorEvent(
+    error: { message: string; stack?: string },
+    additionalData?: any,
```
Comment (lcampos) [inline packages/salesforcedx-vscode-apex/src/telemetry/telemetry.ts]: I think commandName can be sent as part of additionalData.

Also, is it work deleting the `sendError` error method and update the code to use the new `sendErrorEvent` ?

Comment (brpowell) [reply packages/salesforcedx-vscode-apex/src/telemetry/telemetry.ts]: You mean `sendError` in the core module? It has quite a few references, but not a ridiculous amount. I will create separate work for it. I'll modify additionalData though 👍

**Pull request merged**